### ISSUE TITLE
Change working directory when running a cell instead of the whole file

### ISSUE THREAD

**Issue by abudden** [spyder-ide/spyder#8100]
## Problem Description

I have a script that imports a local module and loads some data.  It includes quite a lot of `#%%` markers to allow me to run part of the script and stop before some slow processing takes place.  When I start Spyder, the last loaded files are still present (which is good), but if I click "Run Current Cell", it fails with a `ModuleNotFoundError`.  If I run the whole script (which takes ages), the working directory gets set correctly and I can then run cell-by-cell.  Alternatively, I can manually change the working directory, but I'd prefer it to be set automatically when I run **part of a file** rather than the just when I run the whole file.

### What steps reproduce the problem?

1. Create a text file "test.txt" and a python file in the same directory (not the default Spyder directory).  The text file can contain anything; the python file looks like this:

```python
#%% Set up stuff
import os

assert os.path.exists("test.txt")
print("Okay")

#%% Do something that takes ages
import time
time.sleep(60)
print("Finished")
```

2. Set up Spyder such that in the "Run" settings, the "Default working directory" is "The directory of the file being executed".
3. Hit "Run" - it will correctly print "Okay" and then about 60 seconds later print "Finished"
4. Quit Spyder and restart it
5. Put the cursor somewhere around the assert and hit "Run current cell"

### What is the expected output? What do you see instead?

Expected output is "Okay", actual output is an assertion error.

If you hit "Run", wait 60 seconds and then try "Run current cell" again, it'll work as expected.

## Versions

* Spyder version: 3.3.1
* Python version: 3.6.6 64 bits
* Qt version: 5.9.6
* PyQt version: 5.9.2
* Operating System name/version: Windows 10

### Dependencies

```
IPython >=4.0     :  6.5.0 (OK)
cython >=0.21     :  0.28.5 (OK)
jedi >=0.9.0      :  0.12.1 (OK)
matplotlib >=2.0.0:  2.2.3 (OK)
nbconvert >=4.0   :  5.4.0 (OK)
numpy >=1.7       :  1.15.1 (OK)
pandas >=0.13.1   :  0.23.4 (OK)
pycodestyle >=2.3 :  2.4.0 (OK)
pyflakes >=0.6.0  :  2.0.0 (OK)
pygments >=2.0    :  2.2.0 (OK)
pylint >=0.25     :  2.1.1 (OK)
qtconsole >=4.2.0 :  4.4.1 (OK)
rope >=0.9.4      :  0.11.0 (OK)
sphinx >=0.6.6    :  1.7.9 (OK)
sympy >=0.7.3     :  1.2 (OK)
```

**Comment by ccordoba12** [comment]
Thanks for reporting.

@bcolsen, what do you think about this?

**Comment by bcolsen** [comment]
@abudden This was actually next on my list to fix. I do this all the time.

Here is my current work around: After your data import code that you need to run first put a undefined variable name like "halt" I use "yo" whatever isn't defined in your script. Then run the file using F5. The script will stop at the undefined name with an error, but the data and imports will will stay in the shell. Then you can run the code cells.

**Comment by abudden** [comment]
@bcolsen Thanks, I'll look forward to this fix.

At the minimum, simply setting `__file__` to the path of the file from which the snippet is being run would be sufficient.  I could then add this at the top of my script template:

```
import os
os.chdir(os.path.abspath(os.path.dirname(__file__)))
```

**Comment by bcolsen** [comment]
> @bcolsen Thanks, I'll look forward to this fix.

@abudden Yeah, no problem.

You may also be interested to know that with the new `runcell` in Syder 4 if you `runfile` on a script in a dedicated console then runcell snippets will run only in the console as well.

**Comment by TheUchi** [comment]
I am still experiencing the same bug in Spyder 3.3.6 and the workaround with "halt" is really ugly. The solution proposed by abudden is also not working, since __file__is not defined. Will this bug be fixed in Spyder 3 or do i have to wait until Spyder 4 is stable enough?

**Comment by bcolsen** [comment]
@TheUchi It hasn't been fixed in at all yet but I've been using runfile and sys.exit() lately which is better than some unnamed variable:

```python
from sys import exit as exit

# Do long data import/creation

exit()
# %% Data analysis code cells
```

exit() give this output in the console:
```python
 An exception has occurred, use %tb to see the full traceback.

SystemExit

/home/bcolsen/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3275: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)
```

**Comment by TheUchi** [comment]
@bcolsen Yeah, this looks much better than random variables. Thank you very much!

**Comment by ccordoba12** [comment]
@bcolsen, what do you think about adding `__file__` to `runcell`?

**Comment by bcolsen** [comment]
> what do you think about adding `__file__` to `runcell`?

That should be simple enough, so I don't see why not. I think we should delete `__file__` it after `runcell` execution.

That being said I still feel like `os.chdir(os.path.abspath(os.path.dirname(__file__)))` is a work around as well.

One solution would be having a function that opens a dedicated console that changes to the working directory of the current file in the editor, and maybe a function that just changes the working directory of the current console to the current file as well.

The work flow would be:
1. Open a file with code cells
2. Run the new "open dedicated console" functionality
3. Cells from that file will run in that dedicated console by default 

The new option could be under consoles as "new dedicated console" .

**Comment by ccordoba12** [comment]
> Run the new "open dedicated console" functionality

Are you proposing a new type of dedicated console? Or just the current one?

**Comment by bcolsen** [comment]
Just the current dedicated console through the ipython client's `create_client_for_file`. We would just have to make a new console type like #7099 and run `create_client_for_file` with the current file. I think the working directory might need to be changed as well when the dedicated console is setup. With the new `runcell` function all cells in that file would run in the dedicated console.

**Comment by bcolsen** [comment]
As a note here. I forgot that I already implemented setting __file__ for `runcell` in https://github.com/spyder-ide/spyder-kernels/pull/77

I think we should leave this issue open to provide an IDE option to set the console working directory and/or make a dedicated console.

**Comment by bcolsen** [comment]
Another note here. You can already change the working directory of the current ipython console in the editor by selection the option in the right-click menu of the file tab

**Comment by bcolsen** [comment]
#4344 could be solved with this as well

**Comment by fbosio** [comment]
> @TheUchi It hasn't been fixed in at all yet but I've been using runfile and sys.exit() lately which is better than some unnamed variable:
> 
> ```python
> from sys import exit as exit
> 
> # Do long data import/creation
> 
> exit()
> # %% Data analysis code cells
> ```
> 
> exit() give this output in the console:
> 
> ```python
>  An exception has occurred, use %tb to see the full traceback.
> 
> SystemExit
> 
> /home/bcolsen/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3275: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
>   warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)
> ```

@bcolsen @TheUchi I know this was discussed a year ago, but let me add that you could use `raise` too:
```python
# Do long data import/creation

raise
# %% Data analysis code cells
```

@bcolsen 
> Another note here. You can already change the working directory of the current ipython console in the editor by selection the option in the right-click menu of the file tab

It works great! Thanks!!
